pkg/controllers/disruption/orchestration/queue.go

```diff
@@ -0,0 +1,326 @@
+		nodeClaim, err := nodeclaim.Get(ctx, q.kubeClient, key.Key)
+		if err != nil {
+			// The NodeClaim got deleted after an initial eventual consistency delay
+			// This means that there was an ICE error or the Node initializationTTL expired
+			// In this case, the error is unrecoverable, so don't requeue.
+			if apierrors.IsNotFound(err) && q.clock.Since(cmd.TimeAdded) > time.Second*5 {
+				return NewUnrecoverableError(fmt.Errorf("replacement was deleted, %w", err))
+			}
+			waitErrs[i] = fmt.Errorf("getting node claim, %w", err)
+			continue
+		}
+		// We emitted this event when disruption was blocked on launching/termination.
+		// This does not block other forms of deprovisioning, but we should still emit this.
+		q.recorder.Publish(disruptionevents.Launching(nodeClaim, cmd.Reason()))
+		initializedStatus := nodeClaim.StatusConditions().GetCondition(v1beta1.Initialized)
+		if !initializedStatus.IsTrue() {
+			q.recorder.Publish(disruptionevents.WaitingOnReadiness(nodeClaim))
+			waitErrs[i] = fmt.Errorf("node claim not initialized")
+			continue
+		}
+		cmd.ReplacementKeys[i].Initialized = true
+		// Subtract the last initialization time from the time the command was added to get initialization duration.
+		initLength := initializedStatus.LastTransitionTime.Inner.Time.Sub(nodeClaim.CreationTimestamp.Time).Seconds()
+		disruptionReplacementNodeClaimInitializedHistogram.Observe(initLength)
+	}
+	// If we have any errors, don't continue
+	if err := multierr.Combine(waitErrs...); err != nil {
+		return fmt.Errorf("waiting for replacement initialization, %w", err)
+	}
+
+	// All replacements have been provisioned.
+	// All we need to do now is get a successful delete call for each node claim,
+	// then the termination controller will handle the eventual deletion of the nodes.
+	var multiErr error
+	for i := range cmd.Candidates {
+		candidate := cmd.Candidates[i]
+		q.recorder.Publish(disruptionevents.Terminating(candidate.Node, candidate.NodeClaim, cmd.Reason())...)
+		if err := nodeclaim.Delete(ctx, q.kubeClient, candidate.NodeClaim); err != nil {
+			multiErr = multierr.Append(multiErr, client.IgnoreNotFound(err))
+		} else {
+			nodeclaim.TerminatedCounter(cmd.Candidates[i].NodeClaim, cmd.Method).Inc()
+		}
+	}
+	// If there were any deletion failures, we should requeue.
+	// In the case where we requeue, but the timeout for the command is reached, we'll mark this as a failure.
+	if multiErr != nil {
+		return fmt.Errorf("terminating nodeclaims, %w", multiErr)
+	}
+	return nil
+}
+
+// Add adds commands to the Queue
+// Each command added to the queue should already be validated and ready for execution.
+func (q *Queue) Add(cmd *Command) error {
+	providerIDs := lo.Map(cmd.Candidates, func(s *state.StateNode, _ int) string {
+		return s.ProviderID()
+	})
+	// First check if we can add the command.
+	if err := q.CanAdd(providerIDs...); err != nil {
+		return fmt.Errorf("adding command, %w", err)
+	}
+	q.mu.Lock()
+	defer q.mu.Unlock()
```

Comment:
I'd personally just throw the `q.mu.Unlock` on L292 since defer will wait unnecessarily longer in the function than needed

Reply:
Shouldn't we always be deferring? I'm imagining in the following lines where any of this unexpectedly stops/panics/exits that we should be ensuring that this is unlocked eventually. 

```
	for _, candidate := range cmd.Candidates {
		q.ProviderIDToCommand[candidate.ProviderID()] = cmd
	}
```

Reply:
I don't have a huge issue with the way that it is right now. I think it _might_ be more technically correct, but this matches the existing pattern so it's fine